pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/dictionary/DictionaryIndexType.java

```diff
@@ -107,11 +116,15 @@ public Map<String, DictionaryIndexConfig> fromIndexLoadingConfig(
     Set<String> noDictionaryCols = indexLoadingConfig.getNoDictionaryColumns();
     Set<String> onHeapCols = indexLoadingConfig.getOnHeapDictionaryColumns();
     Set<String> varLengthCols = indexLoadingConfig.getVarLengthDictionaryColumns();
+    Map<String, OnHeapDictionaryConfig> onHeapDictionaryConfigs = indexLoadingConfig.getOnHeapDictionaryConfigs();
     for (String column : indexLoadingConfig.getAllKnownColumns()) {
       if (noDictionaryCols.contains(column)) {
         result.put(column, DictionaryIndexConfig.disabled());
       } else {
-        result.put(column, new DictionaryIndexConfig(onHeapCols.contains(column), varLengthCols.contains(column)));
+        boolean loadOnHeap = onHeapCols.contains(column);
+        OnHeapDictionaryConfig onHeapConfig = loadOnHeap ? onHeapDictionaryConfigs.get(column) : null;
+        result.put(column,
+            new DictionaryIndexConfig(onHeapCols.contains(column), varLengthCols.contains(column), onHeapConfig));
```

Comment:
As said in IndexingConfig comment, I recommend to do not support this feature using the old syntax. By doing so we won't need to modify this part of the code.